### PR TITLE
Add use_cudnn mode

### PR DESCRIPTION
**Pull request by beam2d**
This is a PR that fixes a part of #2048. It introduces `chainer.config.use_cudnn` boolean configuration. When it is True, all functions that support cuDNN use it if possible, and otherwise it uses our own implementations. It replaces the `use_cudnn` argument of many functions so that these APIs are made simpler.

### CODE REVIEW

**Comment by unnonouno** [comment]
Some methods such as softmax function are slower than our implementation when an array is not contiguous. Do we need to make a method to control behavior more precisely?

**Comment by beam2d** [comment]
One explicit way to do it is just switching the mode around the softmax function, like
```
with chainer.using_config('use_cudnn', False):
    y = F.softmax(x)
```
But I feel it is bothersome a bit. Another option is to provide another value for use_cudnn (say, `use_cudnn = 'fastest'`) so that the implementation of each function chooses the best implementation for each function (of course it does not guarantee that the choice is always correct in terms of the actual performance, though).

**Comment by unnonouno** [comment]
How about using options 'always', 'auto', 'never'?

**Comment by beam2d** [comment]
That sounds good. I will use it.

**Comment by beam2d** [comment]
@unnonouno I turned the flag into a three-way option (always, auto, never). Would you check it?

File: chainer/functions/connection/convolution_nd.py [hunk 1]
```diff
@@ -334,7 +330,7 @@ def convolution_nd(x, W, b=None, stride=1, pad=0, use_cudnn=True,
     computation if ALL of the following conditions are satisfied:
 
     - ``cuda.cudnn_enabled`` is ``True``
-    - ``use_cudnn`` is ``True``
+    - ``chainer.config.use_cudnn`` is ``True``
```
Comment (unnonouno) [inline chainer/functions/connection/convolution_nd.py]: `'always' or 'auto'`?

File: tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py [hunk 2]
```diff
@@ -20,7 +20,7 @@ def _split(inputs, pos):
 
 
 @testing.parameterize(*testing.product({
-    'use_cudnn': [True, False],
+    'use_cudnn': ['always', 'never'],
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py]: 'auto'?

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 3]
```diff
@@ -68,8 +68,8 @@ def test_forward_consistency(self):
     @attr.gpu
     @condition.retry(3)
     def test_forward_consistency_im2col(self):
-        self.link.use_cudnn = False
-        self.check_forward_consistency()
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: 'never'?

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 4]
```diff
@@ -89,9 +89,9 @@ def test_backward_gpu(self):
     @attr.gpu
     @condition.retry(3)
     def test_backward_gpu_im2col(self):
-        self.link.use_cudnn = False
         self.link.to_gpu()
-        self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: 'never'?

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 5]
```diff
@@ -167,8 +167,8 @@ def test_forward_consistency(self):
     @attr.gpu
     @condition.retry(3)
     def test_forward_consistency_im2col(self):
-        self.link.use_cudnn = False
-        self.check_forward_consistency()
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: 'never'?

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py [hunk 6]
```diff
@@ -80,8 +80,8 @@ def test_forward_consistency(self):
     @attr.gpu
     @condition.retry(3)
     def test_forward_consistency_im2col(self):
-        self.link.use_cudnn = False
-        self.check_forward_consistency()
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py]: 'never'?

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py [hunk 7]
```diff
@@ -101,9 +101,9 @@ def test_backward_gpu(self):
     @attr.gpu
     @condition.retry(3)
     def test_backward_gpu_im2col(self):
-        self.link.use_cudnn = False
         self.link.to_gpu()
-        self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py]: 'never'?

File: tests/chainer_tests/links_tests/connection_tests/test_deconvolution_2d.py [hunk 8]
```diff
@@ -69,8 +69,8 @@ def check_forward_consistency(self):
     @attr.gpu
     @condition.retry(3)
     def test_forward_consistency(self):
-        self.link.use_cudnn = self.use_cudnn
-        self.check_forward_consistency()
+        with chainer.using_config('use_cudnn', self.use_cudnn):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_deconvolution_2d.py]: set use_cudnn `['always', 'auto', 'never']`

File: tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py [hunk 9]
```diff
@@ -63,8 +63,8 @@ def test_forward_consistency(self):
     @attr.gpu
     @condition.retry(3)
     def test_forward_consistency_im2col(self):
-        self.link.use_cudnn = False
-        self.check_forward_consistency()
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py]: 'never'

File: tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py [hunk 10]
```diff
@@ -83,9 +83,9 @@ def test_backward_gpu(self):
     @attr.gpu
     @condition.retry(3)
     def test_backward_gpu_im2col(self):
-        self.link.use_cudnn = False
         self.link.to_gpu()
-        self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py]: 'never'

File: tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py [hunk 11]
```diff
@@ -164,8 +164,8 @@ def test_forward_consistency(self):
     @attr.gpu
     @condition.retry(3)
     def test_forward_consistency_im2col(self):
-        self.link.use_cudnn = False
-        self.check_forward_consistency()
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py]: 'never'

File: tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py [hunk 12]
```diff
@@ -184,9 +184,9 @@ def test_backward_gpu(self):
     @attr.gpu
     @condition.retry(3)
     def test_backward_gpu_im2col(self):
-        self.link.use_cudnn = False
         self.link.to_gpu()
-        self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+        with chainer.using_config('use_cudnn', False):
```
Comment (unnonouno) [inline tests/chainer_tests/links_tests/connection_tests/test_dilated_convolution_2d.py]: 'never'

File: chainer/functions/pooling/average_pooling_2d.py [hunk 13]
```diff
@@ -65,8 +66,7 @@ def backward_cpu(self, x, gy):
         return gx,
 
     def backward_gpu(self, x, gy):
-        if (cuda.cudnn_enabled and self.use_cudnn and
-                pooling_2d._check_cudnn_acceptable_type(x[0].dtype)):
+        if self._used_cudnn:
```
Comment (unnonouno) [inline chainer/functions/pooling/average_pooling_2d.py]: A user cannot use cudnn only for backward. Is it right?

File: chainer/functions/pooling/pooling_2d.py [hunk 14]
```diff
@@ -48,6 +47,8 @@ def check_type_forward(self, in_types):
         )
 
     def forward_gpu(self, x):
+        self._used_cudnn = True
```
Comment (unnonouno) [inline chainer/functions/pooling/pooling_2d.py]: It always uses cudnn, right? Why is this attribute required?

File: chainer/__init__.py [hunk 15]
```diff
@@ -85,6 +85,50 @@ def get_function_hooks():
...
+_SHOULD_USE_CUDNN = {
+    '==always': {'always': True, 'auto': False, 'never': False},
+    '>=auto':   {'always': True, 'auto': True,  'never': False},
+}
+
+
+_cudnn_version = cuda.cudnn.cudnn.getVersion() if cuda.cudnn_enabled else 1
+
+
+def should_use_cudnn(level, lowest_version=0):
```
Comment (unnonouno) [inline chainer/__init__.py]: Please write unit tests for these methods.

File: chainer/functions/pooling/average_pooling_2d.py (hunk 13 above)
Comment (beam2d) [inline chainer/functions/pooling/average_pooling_2d.py]: Yes, that's right. cuDNN routine requires `self.y` in backward, which is not required and thus retained by CuPy routine, and so we cannot use cuDNN only for backward unless we retain `self.y` in CuPy routine as well.

File: chainer/functions/pooling/pooling_2d.py (hunk 14 above)
Comment (beam2d) [inline chainer/functions/pooling/pooling_2d.py]: Pooling2D is a base class and `forward_gpu` is overridden by each implementation. This function is only called when the implementation chooses cuDNN.

File: chainer/__init__.py (hunk 15 above)
Comment (beam2d) [inline chainer/__init__.py]: Thanks, I'll write.

File: tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py (hunk 2 above)
Comment (beam2d) [inline tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py]: I think we do not need to test `auto` if the test only checks if it runs correctly with and without cuDNN. It reduces the number of test combinations in many cases. Of course, we should still try `auto` if the test checks if the correct mode is chosen with a given `use_cudnn` flag.

File: tests/chainer_tests/links_tests/connection_tests/test_deconvolution_2d.py (hunk 8 above)
Comment (beam2d) [inline tests/chainer_tests/links_tests/connection_tests/test_deconvolution_2d.py]: ditto

**Comment by beam2d** [comment]
@unnonouno I merged the latest _v2 and fixed tests, flake8, and documents.
